### PR TITLE
feat(terser): add source map links

### PR DESCRIPTION
**Pull request by Toxicable**
## PR Checklist
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)

## What is the current behavior?
Issue Number: #1182
minified files do not link back to their source maps

## What is the new behavior?
minified files no link back to the source maps

## Does this PR introduce a breaking change?
- [x] No

### CODE REVIEW

File: packages/terser/src/index.js [hunk 1]
```diff
@@ -48,6 +48,35 @@ log_verbose(`Running terser/index.js
   output: ${output}
   residual: ${residual}`);
 
+/**
+ * @param {string[]} residuleArgs
```
Comment (alexeagle) [inline packages/terser/src/index.js]: spelling

File: packages/terser/src/index.js [hunk 2]
```diff
@@ -48,6 +48,35 @@ log_verbose(`Running terser/index.js
...
+ * @param {string} output
+ */
+function processSourceMapArg(residuleArgs, output) {
```
Comment (alexeagle) [inline packages/terser/src/index.js]: this function seems like it would also benefit from some unit testing?

File: packages/terser/src/index.js [hunk 3]
```diff
@@ -117,6 +148,7 @@ function terserDirectory(input) {
 }
 
 if (!inputs.find(isDirectory) && inputs.length) {
+  process.argv = processSourceMapArg(process.argv, output);
```
Comment (alexeagle) [inline packages/terser/src/index.js]: why mess with the argv if we are just a pass-through to terser? it should be a no-op

File: packages/terser/test/inline_sourcemap/spec.js [hunk 4]
```diff
@@ -14,4 +14,14 @@ describe('terser sourcemap handling', () => {
       expect(pos.column).toBe(14);
     });
   });
+
+  it('should link the source to the source map', () => {
+    const file = require.resolve(__dirname + '/out.min.js');
+    const expectedSourceMapUrl = require.resolve(__dirname + '/out.min.js.map');
+    const content = fs.readFileSync(file, 'utf8');
+    const sourceMapLine = content.split('\n').find(l => l.startsWith('//#'));
```
Comment (alexeagle) [inline packages/terser/test/inline_sourcemap/spec.js]: maybe need to split on `/\r?\n/` for Windows?

File: packages/terser/src/index.js [hunk 5]
```diff
@@ -48,6 +48,35 @@ log_verbose(`Running terser/index.js
...
+  const sourceMapIndex = argsCopy.findIndex(a => a === '--source-map');
+  if (sourceMapIndex != -1) {
+    const sourceMapOptsIndex = sourceMapIndex + 1
+    const sourceMapOpts = argsCopy[sourceMapOptsIndex];
+
+    const baseKeyValue =
+        sourceMapOpts.split(',').map(g => g.split('=')).find(([key, value]) => key === 'base');
+    if (baseKeyValue) {
+      const [_, baseUrl] = baseKeyValue;
+      // TODO: maybe there's a better api to remove this part of the url
```
Comment (alexeagle) [inline packages/terser/src/index.js]: fix the TODO?

File: packages/terser/src/index.js [hunk 6]
```diff
@@ -48,6 +48,35 @@ log_verbose(`Running terser/index.js
...
+      // the .slice(1) remove the starting `/`
+      const relativeUrl = output.replace(baseUrl, '').slice(1);
+      // if we use this relative one them terser throws
```
Comment (alexeagle) [inline packages/terser/src/index.js]: them -> then

File: packages/terser/src/index.js (hunk 3 above)
Comment (Toxicable) [reply packages/terser/src/index.js]: terser reads the args back out of `process.argv` it's the only api we have for it

File: packages/terser/src/index.js (hunk 2 above)
Comment (Toxicable) [reply packages/terser/src/index.js]: it's a little hard to do, should, we'd have to export it from this file to be able to unit test, do you think exporting is a good idea?

Comment (alexeagle) [reply packages/terser/src/index.js]: sure, exporting the function is fine I think, it's not like we need to document it as public API

File: packages/terser/src/index.js (hunk 3 above)
Comment (alexeagle) [reply packages/terser/src/index.js]: yeah I know, but why do you need to do any processing on argv?

Comment (Toxicable) [reply packages/terser/src/index.js]: we need to process the source map argument, in processing we modify it to include the extra option, the same as the directory method

File: packages/terser/src/index.js (hunk 2 above)
Comment (Toxicable) [reply packages/terser/src/index.js]: Aight, added some tests, required a bit of refactoring so needs a rereview

**Comment by Toxicable** [comment]
@alexeagle Looks like we're good on CI, had a few failures, guess they were flakes?

File: packages/terser/test/inline_sourcemap/spec.js [hunk 7]
```diff
@@ -14,4 +15,19 @@ describe('terser sourcemap handling', () => {
       expect(pos.column).toBe(14);
     });
   });
+
+  it('should link the source to the source map', () => {
+    const file = require.resolve(__dirname + '/out.min.js');
```
Comment (soldair) [inline packages/terser/test/inline_sourcemap/spec.js]: Because __dirname is an absolute path `require.resolve` can be replaced with `path.join`
This is helpful because when we look at require.resolve use to refactor out loader.js logic there will be fewer of them to understand.

File: packages/terser/src/terser_minified.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+ * @param {string[]} residualArgs
+ * @param {string} output
+ */
+function processSourceMapArg(residualArgs, output) {
+  // copy args so we don't accidently mutate and process the same set of args twice
+  const argsCopy = [...residualArgs];
+  const sourceMapIndex = argsCopy.findIndex(a => a === '--source-map');
+  if (sourceMapIndex != -1) {
+    // the options for the source map arg is the next one
+    // if it changes in terser_minified.bzl this needs to be updated
```
Comment (soldair) [inline packages/terser/src/terser_minified.js]: we should consider reintroducing minimist for option parsing at the very least to organize which flags we read and why.

Comment (Toxicable) [reply packages/terser/src/terser_minified.js]: i'd appciate having a lib for handling args, doing it manually can get error prone.
Considering almost everthing we handle is CLIs maybe this is a good idea.
Or at very least implement our own standward for CLI args parsing if we dont want to pull in a full lib

File: packages/terser/src/terser_minified.js [hunk 9]
```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env node
```
Comment (alexeagle) [inline packages/terser/src/terser_minified.js]: please rename to the old path, or at least put the rename in its own commit so I don't have a giant block of "newly added" code that I can't tell what you changed

Comment (Toxicable) [reply packages/terser/src/terser_minified.js]: Yeah that's fair, thought the naming of the files made more sense this way, but for the sake of actually able to code review it I'm happy flipping it around

**Comment by Toxicable** [comment]
It's failing on windows with the brlow error, not really sure why
```
1) terser sourcemap handling should link the source to the source map
--
  | Message:
  | Error: ENOENT: no such file or directory, open 'D:\b\bk-windows-java8-kfjp\bazel\rules-nodejs-nodejs\packages\terser\test\inline_sourcemap\out.min.js'
```

File: packages/terser/src/terser_minified.js (hunk 9 above)
Comment (Toxicable) [reply packages/terser/src/terser_minified.js]: put the file rename in a seprate commit

File: packages/terser/src/index.js (hunk 3 above)
Comment (alexeagle) [reply packages/terser/src/index.js]: Sorry I still don't get it. If you're not doing a directory, you don't need our wrapper. The original terser arguments should work as-is

File: packages/terser/test/sourcemap-arg.spec.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
+const {processSourceMapArg} = require('npm_bazel_terser/index')
+
+describe('processSourceMapArg', () => {
+    it('return a new array ref', () => {
+  const args = [];
+  expect(processSourceMapArg(args)).not.toBe(args)
+    })
+
+it('should strip the base and leading / from the output url', () => {
```
Comment (alexeagle) [inline packages/terser/test/sourcemap-arg.spec.js]: why is the formatter not working for you? there should be a husky-installed pre-commit hook to check that you formatted your changes

File: packages/terser/src/index.js [hunk 11]
```diff
@@ -147,3 +150,49 @@ function spawn(cmd, args) {
     });
   })
 }
+
+function main() {
```
Comment (alexeagle) [inline packages/terser/src/index.js]: please don't do unrelated refactorings. this just makes me search around to see what you changed.

Especially as you're on your first few contributions I would like to get minimal deltas where every edit is related to making the new test pass please

File: packages/terser/src/terser_minified.js (hunk 9 above)
Comment (alexeagle) [reply packages/terser/src/terser_minified.js]: that didn't quite work because the first commit still references the new file, I wasn't able to just fetch and get the test failure off the first commit.

I don't even really agree with renaming the file. If there's only ever going to be one file in this package, why not name it index? it makes the rules shorter since there are no data deps

File: packages/terser/src/index.js (hunk 11 above)
Comment (Toxicable) [reply packages/terser/src/index.js]: I needed to do this refactor so that this file wouldn't jsut execute the CLI work flow when required into the tests

File: packages/terser/src/terser_minified.js (hunk 9 above)
Comment (Toxicable) [reply packages/terser/src/terser_minified.js]: > I wasn't able to just fetch and get the test failure off the first commit.

I have a couple of fixups that apply to the first commit, they'd need to be rebased in to allow for that.

I couldn't get `if(require.module === module) { main() }` to work here under the testing environment. It would still call the main method even when that file was being required and wasn't the entry point.
I'd prefer a single file aswell but I don't get how to make it work.

File: packages/terser/src/index.js (hunk 3 above)
Comment (Toxicable) [reply packages/terser/src/index.js]: >The original terser arguments should work as-is

The original terser arguments don't have the `url` set in the sourcemap options, calling this methods sets it.

File: packages/terser/test/sourcemap-arg.spec.js (hunk 10 above)
Comment (Toxicable) [reply packages/terser/test/sourcemap-arg.spec.js]: Im pretty sure thats what the formatter wanted to do.
I even ran it manually, as in `clang-format` instead of `git-clang-format`

Comment (Toxicable) [reply packages/terser/test/sourcemap-arg.spec.js]: ah! it was formatting correctly, but I forgot some semi colons - im use to prettier isnerting them for me

File: packages/terser/src/index.js (hunk 3 above)
Comment (alexeagle) [reply packages/terser/src/index.js]: I see, so then what I'm looking for is that the original terser arguments should include `url` in the sourcemap options so that the original binary works in the single-file case.

File: packages/terser/src/terser_minified.js (hunk 9 above)
Comment (alexeagle) [reply packages/terser/src/terser_minified.js]: you mean `if (require.main === module) {` ?
What didn't work about it? Let's please figure that out rather than introduce a refactoring purely as a workaround for a problem we don't understand

**Comment by alexeagle** [comment]
sorry this is not getting fast enough back and forth - happy to discuss on Slack. This still doesn't look like what I was expecting.

File: packages/terser/src/index.js (hunk 3 above)
Comment (Toxicable) [reply packages/terser/src/index.js]: hmm ok ill see what I can do, but I assume this means we'll be duplicating the logic in the function I added between here and in the `.bzl` that goes with this.

File: packages/terser/src/terser_minified.js (hunk 9 above)
Comment (Toxicable) [reply packages/terser/src/terser_minified.js]: The refactor is still required, if we can get `if (require.main === module) {` to work then that just removed the need for two files.

Comment (Toxicable) [reply packages/terser/src/terser_minified.js]: ok tried it again and it's working now, not sure what happened before

File: packages/terser/src/index.js [hunk 12]
```diff
@@ -22,48 +22,65 @@ function log_error(...m) {
...
-
-log_verbose(`Running terser/index.js
-  inputs: ${inputs}
-  output: ${output}
-  residual: ${residual}`);
+/**
+ * @param {string[]} residualArgs
+ * @param {string} output
+ */
+function processSourceMapArg(residualArgs, output) {
```
Comment (gregmagolan) [inline packages/terser/src/index.js]: a better name so that it is clear that this is for the directory case only would be better I think (maybe just `directoryArgs()`. I would also define this function within terserDirectory as well next to exec() so that its clear what the scope is.

File: packages/terser/src/index.js [hunk 13]
```diff
@@ -22,48 +22,65 @@ function log_error(...m) {
...
+  // copy args so we don't accidently mutate and process the same set of args twice
+  const argsCopy = [...residualArgs];
+  const sourceMapIndex = argsCopy.findIndex(a => a === '--source-map');
```
Comment (gregmagolan) [inline packages/terser/src/index.js]: nit:

a little simpler to use indexOf: `const sourceMapIndex = residualArgs.indexOf('--source-map');`

You can also do this check on `residualArgs` and escape with
```
if (sourceMapIndex === -1)
    return residualArgs;
```
and then make a copy of the args after that.

File: packages/terser/src/index.js [hunk 14]
```diff
@@ -22,48 +22,65 @@ function log_error(...m) {
...
+
   let work = [];
   let active = 0;
   let errors = [];
 
   function exec([inputFile, outputFile]) {
     active++;
-    let args = [TERSER_BINARY, inputFile, '--output', outputFile, ...residual];
+    let args =
+        [terserBinary, inputFile, '--output', outputFile, ...processSourceMapArg(residual, output)];
```
Comment (gregmagolan) [inline packages/terser/src/index.js]: Should be `...processSourceMapArg(residual, outputFile)` right?

File: packages/terser/src/index.js (hunk 12 above)
Comment (gregmagolan) [reply packages/terser/src/index.js]: `directoryArgs()` will also be a good place to check if `${inputFile}.map` exists and if it does then `content=inline` needs to be replaced with `content='${inputFile}.map'`. That can be a follow up PR.

Comment (Toxicable) [reply packages/terser/src/index.js]: we cant define it in `terserDirectory ` unforunately since we need to export it to unit test it

File: packages/terser/src/index.js (hunk 14 above)
Comment (Toxicable) [reply packages/terser/src/index.js]: yes! must have messed that up in a rebase or something

File: packages/terser/src/index.js (hunk 3 above)
Comment (Toxicable) [reply packages/terser/src/index.js]: Turns out no logic is required in the .bzl, we know the exact name.

File: packages/terser/src/index.js (hunk 12 above)
Comment (gregmagolan) [reply packages/terser/src/index.js]: ahhh. yes. that makes sense. the name change is clear enough then :)

File: packages/terser/src/BUILD.bazel [hunk 15]
```diff
@@ -42,6 +42,8 @@ filegroup(
 
 nodejs_binary(
     name = "terser",
-    data = ["@npm//terser"],
+    data = [
```
Comment (gregmagolan) [inline packages/terser/src/BUILD.bazel]: nit: this diff is not needed

File: packages/terser/src/index.js [hunk 16]
```diff
@@ -22,48 +22,50 @@ function log_error(...m) {
...
-  inputs: ${inputs}
-  output: ${output}
-  residual: ${residual}`);
-
 function isDirectory(input) {
   return fs.lstatSync(path.join(process.cwd(), input)).isDirectory();
 }
 
-function terserDirectory(input) {
+/**
```
Comment (gregmagolan) [inline packages/terser/src/index.js]: nit: these param strings are not very useful; better to just add a short description of what this function does IMO

File: packages/terser/src/terser_minified.bzl [hunk 17]
```diff
@@ -138,6 +138,11 @@ def _terser(ctx):
         else:
             fail("When sourcemap is True, there should only be one or none input sourcemaps")
 
+        if len(directory_srcs) == 0:
+            source_map_opts.append("url='%s.js.map'" % ctx.label.name)
+        else:
+            source_map_opts.append("url='<OUTPUT_MAP_FILE>'")
```
Comment (gregmagolan) [inline packages/terser/src/terser_minified.bzl]: a comment that this special key is replaced with the actual filename in the .js file would be good here

File: packages/terser/src/index.js (hunk 16 above)
Comment (Toxicable) [reply packages/terser/src/index.js]: I find them helpful for types, but maybe we just convert this file to TS at some point.

**Comment by alexeagle** [comment]
The magic token `<URL_GOES_HERE>` was both smelly, because you need a comment in one file saying to update along with the other file, and also was breaking some shells that interpret the `<` as a file indirection.

I think it's better just to comma-parse the `source-map` argument and replace the `url` option altogether. Simplifies the `.bzl` code in a way that's symmetric with the output option - in both cases the starlark can be naive about assembling the command-line assuming that the `ctx.label.name` is going to be used everywhere.

I added a commit to make that refactoring.
